### ISSUE TITLE
Partial page load across site collections (Modern team sites

### ISSUE THREAD

**Issue by raymondlittle100** [SharePoint/sp-dev-docs#1971]
#### Category
- [x] Question
- [ ] Typo
- [x] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior

When moving between site collections the page context is refreshed and application customizers are re-loaded with new site details. This used to be the behavior until a couple of weeks ago.

#### Observed Behavior

Until recently this worked as expected but now when you navigate between pages in different site collections you get a partial page load and any application customizers are not re-loaded. 

#### Steps to Reproduce

Below shows a page with an application customizer which passes the current site from the context to a react component. On first load the current site is correct but after clicking a link to a different team site the page is partially refreshed and the application customizer is not re-loaded. This may be by design but it not how this used to work until a couple of weeks ago and I can find nothing from MS describing this change

Original page load
![communicationsites](https://user-images.githubusercontent.com/23637375/41145531-4bb0f340-6af8-11e8-8a64-ac37699f29bb.png)

Component details after clicking link
![financesiteafterlinkclick](https://user-images.githubusercontent.com/23637375/41145543-56046778-6af8-11e8-9136-3cf11d2e69db.png)

**Comment by tdwhite0** [comment]
What you need to do is hook into the navigatedEvent of the app customizer. You can do this in the onInit.
`    
    this.context.application.navigatedEvent.add(this, () => {
       // update your control
    })
`

Another thing I have found is that if you use the url of the page you are on from the browser window.location to make any calls inside webparts you will find that the partial navigation causes their js to run before the url is actually updated. So you should always pass the context around and never read from window.location.

**Comment by raymondlittle100** [comment]
You can use this but the documentation from Microsoft still suggests you use the changedEvent, see documentation below.

This approach worked fine until Microsoft changed the partial page loading to work across site collections whereas previous it was only within the one site collection. All I'm trying to do, and have been for the last few weeks, is get a clear position from them around what is the expected behavior. Its frustrating because all they have to do is say yes its expected or its not. If it is expected here is the correct approach for registering your customizer but there has been nothing 

https://docs.microsoft.com/en-us/sharepoint/dev/spfx/extensions/get-started/using-page-placeholder-with-extensions

**Comment by tdwhite0** [comment]
@raymondlittle100 I agree the partial navigation feature is broken. I am trying to debug issues with it right now in fact. I have the issue where when the page does the partial navigation to a page on a different site collection, it seems to reload components two or more times and causes all their async logic to run multiple times.

**Comment by raymondlittle100** [comment]
@tdwhite0 Thanks for the response good to know its not just me. When i was testing using the changedEvent customizer event my React render methods weren't event getting hit so it was as if the customizer wasn't getting re-loaded.

Slightly disappointed on the lack of response on this one

**Comment by tdwhite0** [comment]
@VesaJuvonen did a recent update for this go out that enabled partial navigation between site collections recently? I recently started experiencing some weird issues and that might explain it.

**Comment by raymondlittle100** [comment]
@tdwhite0  there was a recent change to the customizer events for things like navigating to search application pages and back to ensure the customizer was reloaded but that was a month or two ago. This seemed to start a few weeks ago and I can't find anything to say if its expected or not hence this bug.

**Comment by tdwhite0** [comment]
@raymondlittle100 We also had something that seems related: We rendered <a target="_blank" /> tags to get them to open in new tabs, which just recently broke because SP started taking them over and doing partial navigation.

**Comment by RichG82** [comment]
@tdwhite0 Thanks for the workaround.  I am experiencing the same thing with regard to window.location.  It seems that there is a lot of unexpected behavior when doing a partial navigation even through normal HTML links (missing title sometimes, tooltips text breaking)

@raymondlittle100 same here. App customizers not running again on a partial page navigation is happening for me too.  

I wish we could get a little more traction on this issue.

**Comment by raymondlittle100** [comment]
@tdwhite0  if you're tags are links you can disable the partial page rendering by adding data-interception="off". 

My guess is MS had some code which looks for all links ending with .aspx which used to look for only links pointing to the current site collection and when once of these were hit only reload the content area. This seems to be have been expanded so its not only links in the same team site but any team site.

**Comment by tdwhite0** [comment]
I started a new issue for my specific case: https://github.com/SharePoint/sp-dev-docs/issues/2023

**Comment by raymondlittle100** [comment]
Anything on this?

**Comment by rmillener** [comment]
@raymondlittle100 thanks for the data-interception="off" tip. Our solution was much more convoluted.

Ours is working in Chrome with a this.context.application.navigatedEvent callback but has stopped working in IE11. In IE the customiser (I assume correctly??) disposes and disappears from the page.

This has been absolutely catastrophic for our deployment. I'm surprised this isn't getting more attention.

**Comment by raymondlittle100** [comment]
@VesaJuvonen  any ideas on this?

**Comment by marcDeSantis** [comment]
I am also having this same issue, and glad to know I'm not crazy. We are seeing this with a topbar navigation menu built using AngularJS. As you navigate between site collections, it will often throw the browser back to the originating site collection when navigating through secondary collections.

We just released our intranet on May 30th, and it was working fine up until about 2 weeks ago.

I opened a Premier Support ticket, and they are investigating. Has anybody else done so?

**Comment by marcDeSantis** [comment]
@raymondlittle100 can you explain the data-interception tag? Does that just get adding as an attribute of the anchor tag?

@tdwhite0 do you have an example of how to get the window location from the context?

**Comment by tdwhite0** [comment]
@marcDeSantis It looks like there is a top level routing interceptor that on every click will determine where your link is going and attempt to do a partial navigation on it. For example this will intercept any click on an <a/> tag, which is why target="_blank" stopped working on those. However if you set data-interception="off" on the element, it will get ignored and do the old behavior.

For getting the url I make sure to always use the passed in this.context.pageContext.web.absoluteUrl instead of reading it out of window.location.

**Comment by raymondlittle100** [comment]
SharePoint seems by default to intercept any link pointing to the target tenant and only do a partial page load. This used to be only the current site collection and seems to be where the change has been.

By adding this data attribute to any link it stops the partial page load and does a full refresh

**Comment by marcDeSantis** [comment]
This data-interception attribute is undocumented, I assume?

**Comment by raymondlittle100** [comment]
It is supposed because it's what Microsoft do to links set via content editing which are flagged as open in new tab

**Comment by marcDeSantis** [comment]
I have verified that data-interception="off" has also fixed my issue. Am still curious what Microsoft's recommendation is here, and whether they will take these application extensions into account before making breaking changes down the road.

**Comment by raymondlittle100** [comment]
Glad to hear it. It's frustrating at the lack of progress or comments

**Comment by rmillener** [comment]
If its any consolation it appears that the OOTB quick launch is also affected as per #2095 

The simplest case is:
1. Create two modern Team Sites
2. Add a link from one to the other to the quick launch
3. Click the link
4. Site navigates but the quick launch stays the same 

@marcDeSantis did your premier support ticket get anywhere?

**Comment by marcDeSantis** [comment]
> @marcDeSantis did your premier support ticket get anywhere?

Nope. I told them that the `data-interception` workaround was working for me, so they recommended putting in a suggested change to the docs, and mentioned that they cannot guarantee the functionality of customizations 😞

**Comment by raymondlittle100** [comment]
Thanks. Appreciate the update

**Comment by raymondlittle100** [comment]
Any further update on this?

**Comment by raymondlittle100** [comment]
@rmillener  is this still working for you I still get the error?

**Comment by rmillener** [comment]
@raymondlittle100 not last time I checked. We will continue with data-interception until we see an official response around navigatedEvent.

**Comment by raymondlittle100** [comment]
@VesaJuvonen  any more details on this?

**Comment by MarcoSimon** [comment]
@VesaJuvonen is there any news on this? I'm still getting this bug!

**Comment by petercarson** [comment]
@VesaJuvonen also keenly interested in this issue.  We have a client about to launch a modern Intranet that this is a blocker for.

**Comment by marcDeSantis** [comment]
Another side effect of this bug just recently came to my attention. I am not sure if it has always been an issue, or was recently introduced.

**Event Web Part Issue**
When clicking "Add Event" on the Event Web Part within a page that uses the Application Customizer, the browser navigates to a new URL `/_layouts/15/Event.aspx?ListGuid=<GUID>&Mode=Edit`

Now, once you click on the Event creation form, the URL jumps back to the root Site Collection. When you click on "Create Event", an error is thrown indicating `This link is not valid. Check the link and try again.`

@VesaJuvonen today is the 6 month anniversary of this issue. Can we please get an update so that we can understand what we are supposed to do? This is extremely frustrating.

**Comment by westleyMS** [comment]
@marcDeSantis What are you doing in your OnInit() for your extension? If your extension is not aware of the partial page loads, make sure in OnInit() to only register for the event, for placeholders-
this.context.placeholderProvider.changedEvent.add(this, this._renderPlaceHolders);
for script extensions-
this.context.application.navigatedEvent.add(this, this._runExtension);

There is an issue where these events fire more than once, but that is almost fixed, and you can add code to de-bounce the event. See https://github.com/SharePoint/sp-dev-docs/issues/1871

**Comment by marcDeSantis** [comment]
Yes - I am using the changedEvent:

```
  public onInit(): Promise<void> {
    // Added to handle possible changes on the existence of placeholders.
    this.context.placeholderProvider.changedEvent.add(this, this.renderTopbar)

    this.renderTopbar()
    return Promise.resolve<void>()
  }
```

**Comment by patmill** [comment]
Sorry - let me summarize what is happening behind the scenes, what you _should_ be hooked into, and then you can say if your behaviour is different.

First - partial navigation is a whole lot faster, so we would prefer to have everything working correctly, and not messing with the data-interception property.
Second - The goal for extensions is to be loaded once, and kept "alive" for as long as possible.  OnInit is called when the extension is loaded (onDispose when it is unloaded).  If you have the same extension on multiple site collections / lists / etc. then the design is that onInit is _not_ called when you do a partial navigate between them.  This keeps them working fast.

If you look closely, the OnChange notification is hooked up to the placeholderProvider. If the placeholders change (which placeholders exist on the current page, if they are visible or not, etc.) then this event will fire.  It _shouldn't_ fire on a navigation event, unless the target of the navigation has different placeholders.

To capture navigate events, you'll want to hook into the navigatedEvent, as @westleyMS writes.  There is a bug (fixed and rolling out) where the navigated event would fire multiple times.  It should only fire once per navigation event.

Given this information, let me know if things aren't working.  Thanks.

**Comment by marcDeSantis** [comment]
I created a scaled down AppCustomizer and sent it to @westleyMS.

**Comment by raymondlittle100** [comment]
I'm not at the same organisation so am no longer in a position to test in the same way.

**Comment by VelinGeorgiev** [comment]
@patmill , I still have issues with that event. I can tweak the code to make it work for me, but they are still there.

Here are the three things still an issue for me:
- `OnInit` of the customizer fires 2 times. There seem to be two instances of my customizer class instead of one.
- `this.context.application.navigatedEvent.add` sometimes fires 2 times instead of 1.
- `this.context.application.navigatedEvent.add` does not fire after navigated when page is not cached in IndexedDB, but before.

I will explain in details bellow with the hope that it will help @patmill, but also the community to workaround that for now. 

My SPFx version is 1.7.0
The assembly I have on my tenant is the following
![hits_assembly](https://user-images.githubusercontent.com/1049476/52952251-e8214800-337b-11e9-9d47-d86cc75bec92.PNG)

**Issue #1 - OnInit of the app customizer fires two times. There seem to be two instances of my app customzier class** instantiated from the internal SharePoint redux single page application router SharePoint team is building. 

I have my app customizer installed only once for the communication site
![onecustomactioninstalled](https://user-images.githubusercontent.com/1049476/52952427-6382f980-337c-11e9-9a62-c6f8669e0875.PNG)


, but when I add console log on the onInit method of the app costomizer, the onInit is being called twice and **NOT from the same instance of the class**. There seem to be two app customizer classes running simultaneously for some reason.


![hits_oninit](https://user-images.githubusercontent.com/1049476/52952496-9fb65a00-337c-11e9-8d6b-0e7a7ac3ea38.PNG)

The problem with that is that when I subscribe to `this.context.application.navigatedEvent.add` from `onInit`, I will subscribe twice if I do not put extra checks to prevent that. 
 
**Issue #2 - `this.context.application.navigatedEvent.add` sometimes fires twice or four times in total if `onInit` fires twice as well and I've subscribed from there**

![untitled1](https://user-images.githubusercontent.com/1049476/52952852-c7f28880-337d-11e9-9e09-37b4be8d05fa.gif)

Here is what happens on the GIF. The `this.context.application.navigatedEvent.add` fires once, waits for some moments and fires again. Combined with the fact that `onInit` fires twice, I get `navigatedEvent` fired 4 times.

**Issue #3 - `this.context.application.navigatedEvent.add` does not fire after navigated when page is not cached in IndexedDB.**

So the description of the `navigatedEvent` says that is fires after navigated, but it is wrong in specific cases
![hits_after](https://user-images.githubusercontent.com/1049476/52953117-73034200-337e-11e9-91e9-37f53508eafe.PNG)


I noticed that is fires relatively after page load only when the page is pulled from the client (IndexedDB). If the Microsoft router has to fetch the page from the server, then the `navigatedEvent` event is fired before the page transition. 

**3 Tweaks to make it work**

Here are the 3 tweaks I had to make in my code so that event fires only once and after the page has loaded. Of-course would be great if I do not have to do that in future. 

Here is my code to make it work:
```
export default class CustomizerLifecycleEventsApplicationCustomizer
  extends BaseApplicationCustomizer<ICustomizerLifecycleEventsApplicationCustomizerProperties> {

  @override
  public onInit(): Promise<void> {

    console.log(`LCEVENT:onInit=${window.location.href}`);

    if (!(window as any).isNavigatedEventSubscribed) {

      this.context.application.navigatedEvent.add(this, this.logNavigatedEvent);

      (window as any).isNavigatedEventSubscribed = true;
    }

    return Promise.resolve();
  }

  @override
  public onDispose(): Promise<void> {

    console.log(`LCEVENT:onDispose=${window.location.href}`);

    this.context.application.navigatedEvent.remove(this, this.logNavigatedEvent);

    (window as any).isNavigatedEventSubscribed = false;
    (window as any).currentPage = '';

    return Promise.resolve();
  }

  public logNavigatedEvent(args: SPEventArgs): void {

    setTimeout(() => {

      if ((window as any).currentPage !== window.location.href) {

        // REGISTER PAGE VIEW HERE >>>
        console.log(`LCEVENT:navigatedEvent=${window.location.href}`);

        (window as any).currentPage = window.location.href;
      }
    }, 3000);
  }
}
```

So you can see some ugly stuff happening, but this is the way to make it work.
- Issue #1. Subscribing twice is solved by having  `window.isNavigatedEventSubscribed` flag to indicate that a subscription is already set before.
- Issue #2. Firing twice `navigatedEvent` is solved by having another `window.currentPage` flag to indicate that this is actually the same `navigatedEvent` and we do not have to do anything.
- Issue #3. Having `setTimeout` for 3000 milliseconds is likely to fix the timing issues and ensure that my code will start executing after the new page is loaded by the client router. I can increase or decrease this time.

**Comment by westleyMS** [comment]
the double firing event has been resolved. My tenant was fixed, and that rollout should be complete by today.

**Comment by Waseemnaseem** [comment]
Good pro

**Comment by andrewconnell** [comment]
Appears this issue is fixed from [this comment](https://github.com/SharePoint/sp-dev-docs/issues/1971#issuecomment-467938875)... if not, we can reopen the issue.

**Comment by tknaller** [comment]
Apparantly partial page loads are now also happening when switching from classic view to modern view...when the happens (not always, some links are not intercepted link page contents) the application customizer doesn't get loaded at all (no onInit trigger) 

apparantly SP thinks that the placeholders have been there before the navigate event even though it was a classic experience before and the placeholders have never been initialized...
